platforms/cuda/src/CudaTorchKernels.cpp

```diff
@@ -145,12 +151,17 @@ std::vector<torch::jit::IValue> CudaCalcTorchForceKernel::prepareTorchInputs(Con
         CHECK_RESULT(cuCtxSynchronize(), "Failed to synchronize the CUDA context"); // Synchronize before switching to the PyTorch context
     }
     // Prepare the input of the PyTorch model
-    vector<torch::jit::IValue> inputs = {posTensor};
+    inputs = {posTensor};
     if (usePeriodic)
         inputs.push_back(boxTensor);
-    for (const string& name : globalNames)
-        inputs.push_back(torch::tensor(context.getParameter(name)));
-    return inputs;
+    for (const string& name : globalNames) {
+        // PyTorch requires us to set requires_grad to false before initializing a tensor.
+        globalTensors[name].set_requires_grad(false);
+        globalTensors[name][0] = context.getParameter(name);
+        if (paramDerivs.find(name) != paramDerivs.end())
+            globalTensors[name].set_requires_grad(true);
+        inputs.push_back(globalTensors[name]);
+    }
```

Comment:
I reckon you are resetting requires_grad as a replacement to zeroing the grads. I am surprised this works with CUDA graphs, since by doing this pytorch might be somehow freeing the gradient tensors and picking up new ones when reenabling.
I am weary you are depending on some UB by doing so.

Reply:
I zero the grads at the end of `execute()`, just before returning.  This is because assigning a value to a tensor throws an exception if you don't set `require_grads` to false first.